skrobot/sdf/signed_distance_function.py

```diff
@@ -0,0 +1,398 @@
+
+        Parameters
+        ----------
+        n_sample : int
+            number of sample points.
+
+        Returns
+        -------
+        :obj: `tuple` of numpy.ndarray[float](n_point, 3),
+        numpy.ndarray[float](n_point,)
+            sampled points and corresponding signed distances.
+        """
+        points_, dists = self._surface_points(n_sample=n_sample)
+        points_obj = self._transform_pts_sdf_to_obj(points_)
+        return points_obj, dists
+
+    def _transform_pts_obj_to_sdf(self, points_obj):
+        """Transform points from to an object coordinate to a sdf-specific coordinate.
+
+        Parameters
+        ----------
+        points_obj : numpy.ndarray[float](n_point x 3)
+            2 dim point array w.r.t. an object coordinate.
+
+        Returns
+        -------
+        points_sdf : numpy.ndarray[float](n_point x 3)
+            2 dim point array w.r.t. a sdf-specific coordinate.
+        """
+        points_sdf = self.coords.copy_worldcoords().transform(
+            self.sdf_to_obj_transform).inverse_transform_vector(points_obj)
+        return points_sdf
+
+    def _transform_pts_sdf_to_obj(self, points_sdf):
+        """Transform points from to a sdf-specific coordinate to an object coordinate.
+
+        Parameters
+        ----------
+        points_sdf : numpy.ndarray[float](n_point x 3)
+            2 dim point array w.r.t. a sdf-specific coordinate.
+
+        Returns
+        -------
+        points_obj : numpy.ndarray[float](n_point x 3)
+            2 dim point array w.r.t. an object coordinate.
+        """
+        points_obj = self.coords.copy_worldcoords().transform(
+            self.sdf_to_obj_transform).transform_vector(
+                points_sdf.astype(np.float32))
+        return points_obj
+
+
+class UnionSDF(SignedDistanceFunction):
+    """One can concat multiple SDFs `sdf_list` by using this class.
+
+    For consistency in the concatenation, it is required that
+    the all SDFs to be concated are with `use_abs=False`.
+    """
+
+    def __init__(self, sdf_list, coords=None):
+        origin = np.zeros(3)
+        use_abs = False
+        super(UnionSDF, self).__init__(origin, coords=coords, use_abs=use_abs)
+
+        use_abs_list = [sdf.use_abs for sdf in sdf_list]
+        all_false = np.all(~np.array(use_abs_list))
+        assert all_false, "use_abs for each sdf must be consistent"
+
+        self.sdf_list = sdf_list
+
+        threshold_list = [sdf._surface_threshold for sdf in sdf_list]
+        self._surface_threshold = max(threshold_list)
+
+    def _signed_distance(self, points_sdf):
+        sd_vals_list = np.array([sdf(points_sdf) for sdf in self.sdf_list])
+        sd_vals_union = np.min(sd_vals_list, axis=0)
+        return sd_vals_union
+
+    def _surface_points(self, n_sample=1000):
+        # equaly asign sample number to each sdf.surface_points()
+        n_list = len(self.sdf_list)
+        n_sample_each = int(floor(n_sample / n_list))
```

Comment:
Please add "from __future__ import division" at the top of this script.

Reply:
ok